Makefile overhaul

magnumripper opened the pull request:
Add real auto-dependency stuff (gcc's own) and drop the incompetent perl script we had for plugins only. The old support for "makedepend" is kept for non-gcc systems.

Fix bugs in format's plugin stanzas.

Have "make distclean" also clear kernel caches (the ones within our tree only).

Move plugin logic back from configure to Makefile. Having them in configure was a major brainfart. Also enhance it so we can actually move plugins (formats or not) into or out of the source tree and just "make" regardless of their timestamps. I made the logic very quick for the case of nothing changed - blindly running the preprocessor over 470 files takes time so we do not want that unless actually needed.
For the same reason, I did not add the plugins themselves as dependencies for the "format register" stuff in john.c. That would mean very slow handling of repeated edit-make with a specific plugin and it's only needed if we do things like change the format struct name.

See #3566 and #3842.

magnumripper committed:
Makefile: Auto dependencies

Only active with gcc/clang, for portability.

See #3566, #3842

magnumripper committed:
Fix bugs in a few formats' plugin logic

Some formats had an #elif or #else creatively skipped in the plugin logic,
messing things up royally and slowing down the build process. It's not a
good idea to write that stuff in any other way, or put anything before it
(with very few exceptions, one being an outermost #if HAVE_OPENCL).

magnumripper committed:
Makefile: Have "make distclean" also clear kernel caches

In this case we do not clear nvidia's caches, just the ones in our tree.

magnumripper committed:
configure/make: Improve plugin handling

Move the plugin logic back to Makefile where it belongs, so we can add or
remove plugins without re-running configure.  Add logic for fast detection
of added or removed plugins.

See #3566

solardiz commented:
Thanks. This is a lot. I don't expect to have time to review it in the next 2 weeks.

magnumripper committed:
Change sm3_plug.c to a non-plugin

It's needed for the dynamic format. We're supposed to be able to delete
every single plugin and build should still succeed.

magnumripper committed:
Autoconf: Test compiler support for auto dependency generation

Instead of disabling dependency generation for any compiler not claiming
to be gcc compatible (but that may support the options anyway), test if
the compiler supports the flags using our existing flag-check wizardry.

magnumripper commented:
Notes to self as well as for your consideration, here's what I have tested:
- Configure + build after `make distclean`.
- Configure with OpenCL disabled + build.

With no re-configure:
- make with no changes.
- make after touching a plugin.
- make after adding a plugin.
- make after removing a plugin.
- make after touching `color.h`.
- make after touching `params.h`.
- make after changing git HEAD.
- <s>make emulating a non-gcc-compatible compiler (auto dependency generation are turned off).</s> (obsolete, see below)
- make after deleting ALL plugin files. I had to add a commit promoting sm3_plug.c to a non-plug because the dynamic format needs it present, but that would have been needed even without this entire PR.
- make after putting all the plugin files back.

Instead of simply disabling auto dependency generation for compilers that MAY be incompatible, we could have the configure script detect if the needed options are supported. We do that for several other things. I'll give that a try and possibly add it as a separate commit.

magnumripper commented:
> Instead of simply disabling auto dependency generation for compilers that MAY be incompatible, we could have the configure script detect if the needed options are supported. We do that for several other things. I'll give that a try and possibly add it as a separate commit.

Done, trivial, works like a champ.

magnumripper committed:
Makefile: Autodependency tweaks

Drop more old fixed dependencies. Most of these were added 10 years ago
and generated by the naive perl script so not great even then.

Add a few dependencies for $(DEPDIR) so we can build particular stuff
from a distclean source directory (i.e. with .deps directory missing).
An example - that a buildbot found - is "make unit-tests".

Also merely moves the symlink recipe for rar2john a little bit just for
easier side to side compare with Makefile.legacy.

solardiz commented:
I still have no time for a proper review, but:

I successfully built from this PR's branch in a fresh Docker container with Alpine Linux (so with musl) as a non-root user by `./configure --without-openssl` and `make -sj2`, after having run as root `apk add make gcc git libc-dev linux-headers`.

> Change sm3_plug.c to a non-plugin
>
> It's needed for the dynamic format. We're supposed to be able to delete
> every single plugin and build should still succeed.

Hmm, but doesn't our CircleCI no-plugs job test that? If so, why is it succeeding before these changes?

magnumripper commented:
> Hmm, but doesn't our CircleCI no-plugs job test that? If so, why is it succeeding before these changes?

I didn't look now but possibly it only deletes plugin _formats_, not other files. Or maybe there's just an exception but then we should have done this earlier.

Now, when I do delete all plugin sources, there will be two "errors" from building (reason being we preprocess using `gcc -E *_plug.c` resulting in "no input files") . While they do stick out they aren't fatal: The build continues and succeeds. I left that as-is but it can be avoided if we want to. OTOH I think such a build is extremely rare, not sure why anyone would do it except for testing the plugin gearing.

magnumripper commented:
> This is a lot.

It isn't really, especially if you review commit for commit. But take your time.

On another note, I realise we could make the whole dynamic format beast plugin(s) now without much effort (there are lots of `#if DYNAMIC_DISABLED` left from a period when it was defunct and Makefile could define that unless the plugs are present). Could be a separate PR though, if we want it at all.

magnumripper commented:
Oh BTW I also tried with gcc 3.3.6 but that fails for other reasons than this PR.
```
argon2_core_plug.c: In function `argon2_validate_inputs':
argon2_core_plug.c:478: warning: comparison is always false due to limited range of data type
blake2b_plug.c: In function `blake2b_compress':
blake2b_plug.c:197: warning: dereferencing type-punned pointer will break strict-aliasing rules
gost3411-2012-core_plug.c: In function `g':
gost3411-2012-core_plug.c:126: warning: implicit declaration of function `_mm_cvtsi64_si128'
gost3411-2012-core_plug.c:126: error: incompatible types in assignment
```

EDIT: See #4271

Maybe we should fix that but I'm not sure I'm interested myself. Also, according to ChatGPT gcc has supported the flags needed for dependency generation since 3.0 but it may well be hallucinating.

magnumripper commented:
Added another commit, for building particular targets (such as unit-tests, found by a buildbot) from a distclean source directory (i.e. with `.deps/` missing).

magnumripper committed:
Makefile: Minor shuffle

Drop the redundant CFLAGSX variable.

Move @DEPFLAGS@ later, after filename and other options. I use to pipe
make's output through a truncating tool of sorts, and the other options
are more useful to have kept in that output.

magnumripper commented:
Added another little commit that just moves @DEPFLAGS@ later on the command line (they are among the least interesting things to see) and drops the redundant CFLAGSX. Not sure why we ever had the latter, it was just confusing.

Rebased on current bleeding while at it.

magnumripper commented:
> > This is a lot.
> 
> It isn't really, especially if you review commit for commit.

@solardiz with 2 months lead time for reviewing a PR that doesn't even change any functionality in john, my incentive for contributing plummets. We could merge this and if you find some flaw later we can fix it then.

solardiz commented:
@magnumripper I'm really sorry. Yes, you may just merge this. I also hope to take a look soon, whether before or after merging. Thank you!

solardiz commented on src/configure.ac:
```diff
@@ -227,6 +226,9 @@ if test "x$ubsantrap" = xyes ; then
 
+   dnl Enable auto dependency generation if supported by the compiler
+   JTR_FLAG_CHECK([-MMD -MP -MF /tmp/.dep], 1)
+   AC_SUBST([DEPFLAGS],['-MMD -MP -MF $(DEPDIR)/$(notdir $*).d'])
```
I think the usage of `/tmp/.dep` in this test is unacceptable - this is basically a vulnerability. Can we use `/dev/null` or a file in the current directory?

Regardless, maybe we should test not only that the flags don't produce an error, but that output is sane?

solardiz replied:
Maybe we should revert to or add a check for gcc/clang? Like options are not rejected _and_ the compiler is gcc/clang.

magnumripper replied:
That use of /tmp was embarassing - I now use conftest.d (in cwd, as in src). Somehow it's gone after the tests without me deleting it. Not sure if I should add a `rm -f conftest.d` anyway?

> Regardless, maybe we should test not only that the flags don't produce an error, but that output is sane?

Done. Our "flag check" verifies that unsupported options do fail though, and if not we'll neither test nor blindly use these options.

magnumripper replied:
> Somehow it's gone after the tests without me deleting it. Not sure if I should add a `rm -f conftest.d` anyway?

I see now the configure script takes care of it due to its name. All build tests in autoconf use the name conftest.c and in the end it deletes conftest.* among other known build artefacts. So we're fine as-is.

solardiz reviewed (changes requested):
I went through the commits one by one. I like how we're dropping the old Perl-based stuff and over 400 lines. There are many things here I didn't review thoroughly, or would need to read up on GNU make to review fully. But if this works, it works. The only blocker I think is the one misuse of `/tmp`, which we must avoid.

solardiz commented:
> Oh BTW I also tried with gcc 3.3.6 but that fails for other reasons than this PR.

I actually wonder if we're maybe introducing a dependency on a newer version of GNU Make than we required before.
